modules/git/url/url.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package url
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+)
+
+// URL represents a git remote URL
+type URL struct {
+	*url.URL
+	extraMark int // 0 no extra 1 scp 2 file path with no prefix
+}
+
+// String returns the URL's string
+func (u *URL) String() string {
+	switch u.extraMark {
+	case 0:
+		return u.String()
+	case 1:
+		return fmt.Sprintf("%s@%s:%s", u.User.Username(), u.Host, u.Path)
+	case 2:
+		return u.Path
+	default:
+		return ""
+	}
+}
+
+var scpSyntaxRe = regexp.MustCompile(`^([a-zA-Z0-9_]+)@([a-zA-Z0-9._-]+):(.*)$`)
+
+// Parse parse all kinds of git remote URL
+func Parse(remote string) (*URL, error) {
+	u, err := url.Parse(remote)
+	if err == nil {
+		extraMark := 0
+		if u.Scheme == "" && u.Path != "" {
+			u.Scheme = "file"
+			extraMark = 2
+		}
+		return &URL{URL: u, extraMark: extraMark}, nil
+	}
+
+	if !strings.Contains(err.Error(), "first path segment in URL cannot contain colon") {
```

Comment:
Does this case really help the parsing? If it doesn't help, maybe we can skip the hard-coded string check.

Reply:
It's really help when parsing `git@gitea.com:gitea/gitea.git` .

Reply:
> It's really help when parsing `git@gitea.com:gitea/gitea.git` .

I know, but could we just skip this check and let the code below to continue to parse? Then we do not need to hard code the magic string.

Reply:
At least, we can check `@` and `://` first, if the URL string doesn't contain `://` but contains `@`, then we do not use `url.Parse`, then we do not need to hard code any error string.

Indeed we can predict all cases:

1. A string contains `://`: use `url.Parse`
2. else if it contains `@`: use SSH parse
3. else check if it is a local path

Then everything is under control.

Reply:
And filepath may also contains `@`.

Reply:
Yup, so there is the step 3 `else` if step 2 fails.

Reply:
Although I would usually agree that hard coding the error seems distasteful it is the correct thing to do. This is the error string that go reports.

I would say that we should use strings.HasSuffix(": first path ...")  with the colon and the space, or use

```go
	urlBadPathError := errors.New("first path segment in URL cannot contain colon")

	if err != nil && errors.As(err, &urlBadPathError) {
```

Reply:
Code outdated.